core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -109,73 +109,91 @@ class TransactionCoordinator(brokerId: Int,
         txnManager.getTransactionState(transactionalId) match {
           case None =>
             val pid = pidManager.nextPid()
-            val newMetadata: TransactionMetadata = new TransactionMetadata(pid = pid,
+            val newMetadata: TransactionMetadata = new TransactionMetadata(producerId = pid,
               producerEpoch = 0,
               txnTimeoutMs = transactionTimeoutMs,
               state = Empty,
               topicPartitions = collection.mutable.Set.empty[TopicPartition],
-              lastUpdateTimestamp = time.milliseconds())
+              txnLastUpdateTimestamp = time.milliseconds())
 
-            val metadata = txnManager.addTransaction(transactionalId, newMetadata)
+            val epochAndMetadata = txnManager.addTransaction(transactionalId, newMetadata)
 
             // there might be a concurrent thread that has just updated the mapping
             // with the transactional id at the same time; in this case we will
             // treat it as the metadata has existed and update it accordingly
-            metadata synchronized {
-              if (!metadata.eq(newMetadata))
-                initPidWithExistingMetadata(transactionalId, transactionTimeoutMs, responseCallback, metadata)
+            epochAndMetadata synchronized {
```

Comment:
Yeah that is a good question, ideally we should sycnrhonize on `txnMetadata` only, since coordinator epoch will only likely be changed during loading / removal of the metadata. Will refactor on this part.